### ISSUE TITLE
Desktop support (Linux, Windows)

### ISSUE THREAD

**Issue by proninyaroslav** [MaikuB/flutter_local_notifications#746]
I see that you are trying to implement macOS support and it would be appropriate to add Windows and Linux as well. Windows has a native Win32 API, while Linux uses `libnotify` library, which covers cases for different desktop environments.

**Comment by MaikuB** [comment]
As the plugin implements a federated plugin structure, the ability to add support for other platforms is open to the community to do so. It sounds like you're interested in leveraging that work, if so then I would suggest you contribute back by adding support for the other platforms.

**Comment by proninyaroslav** [comment]
Yes, I'll try to help you as soon as I get free time.

**Comment by MaikuB** [comment]
Note that I'm not familiar with either Win32 or Linux. As far as I know, they require using C++, which I'm not particularly familiar with so I'm not well-positioned to implement this myself, though I could probably scrap bits of knowledge here and there to review code if needed

**Comment by proninyaroslav** [comment]
In general, any language that supports FFI can be used. But C++ is probably the easiest option.

**Comment by proninyaroslav** [comment]
Although it might be worth using Dart + FFI.

**Comment by artemsivcev** [comment]
Hi! Our team want to help with this issue. Is there any new information about this feature?

**Comment by MaikuB** [comment]
If you want to help then the best way is to submit a PR. There's a draft one for Linux support https://github.com/MaikuB/flutter_local_notifications/pull/888 that you help the author with, help test out etc

**Comment by MaximRyabovol** [comment]
We have some decision for windows notifications. It uses a [process_run](https://pub.dev/packages/process_run) plugin and powershell commands. It works perfectly in debug, but in realise version of the app there are some problems. If you interesting you can find our variant [here](https://github.com/flutter/flutter/issues/70926).

**Comment by HrX03** [comment]
For linux i found a lib using gnome/gtk notification model https://github.com/canonical/desktop_notifications.dart, might want to take a peek. Saw a pr that already tried to use this model but had some issues iirc, maybe this one could work better.

**Comment by jinfagang** [comment]
Why close? This **still an issue!!**

**Comment by MaikuB** [comment]
@jinfagang this wasn't closed and has a needs help label even so as mentioned before in #1183, support for this would need a PRs/contributions from the community

**Comment by Sunbreak** [comment]
A PoC of notification plugin on Android/iOS/Windows/Linux/macOS: https://github.com/woodemi/quick_notify

**Comment by proninyaroslav** [comment]
I'll be taking over Linux (and possibly Windows) support shortly. Thanks @Sunbreak, I will use your example implementation as a starting point to adapt the `flutter_local_notifications` API.

**Comment by proninyaroslav** [comment]
Started on working.

**Comment by proninyaroslav** [comment]
Duplicated from https://github.com/MaikuB/flutter_local_notifications/pull/1208#issuecomment-868767764:

At the moment I'm planning on using `libnotify` and C++. But there is an option to use [DBus](https://pub.dev/packages/dbus) package, as is done in this project: https://github.com/canonical/desktop_notifications.dart. This will allow to write the Linux part entirely in Dart. What do you think about it? 
Similarly, we can use [win32](https://pub.dev/packages/win32) package for notifications, without using native C++ code, only Dart.

**Comment by Sunbreak** [comment]
> At the moment I'm planning on using `libnotify` and C++. But there is an option to use [DBus](https://pub.dev/packages/dbus) package, as is done in this project: https://github.com/canonical/desktop_notifications.dart. This will allow to write the Linux part entirely in Dart. What do you think about it?

https://github.com/canonical/desktop_notifications.dart is great and it is maintained by Canonical, the company of Ubuntu

> Similarly, we can use [win32](https://pub.dev/packages/win32) package for notifications, without using native C++ code, only Dart.

There're still something to implement within `win32` in order to call `ToastNotificationManager`: https://github.com/timsneath/win32/pull/246

**Comment by proninyaroslav** [comment]
@Sunbreak 
>https://github.com/canonical/desktop_notifications.dart is great and it is maintained by Canonical, the company of Ubuntu

Do you mean to use it inside `flutter_local_notifications`?

**Comment by Sunbreak** [comment]
@proninyaroslav you could take advantage of `desktop_notifications` to implement `flutter_local_notifications`

**Comment by proninyaroslav** [comment]
@Sunbreak 
I don't think this is a good idea as `flutter_local_notifications` tries to avoid external dependencies. It might be worth considering [dbus](https://pub.dev/packages/dbus) package, or implementing DBus calls in native code if [dbus](https://pub.dev/packages/dbus) package is not suitable in this case. @MaikuB what do you think about this?

**Comment by MaikuB** [comment]
@proninyaroslav agreed, the only dependency this plugin should take is to use `dbus` to make calls to show notifications or consume a plugin that implements Linux support by implementing the platform interface. Using `desktop_notifications` would make maintenance more difficult

**Comment by proninyaroslav** [comment]
Linux support implementation is ready, I invite everyone to take part in the review and testing https://github.com/MaikuB/flutter_local_notifications/pull/1208

**Comment by 12people** [comment]
The Linux implementation was swiftly removed again in v 7.0 because of a conflict with the web builds.

Is this going to be a trivial issue to fix or should I not expect Linux notification support anytime soon?

**Comment by MaikuB** [comment]
There's an open PR on this already https://github.com/MaikuB/flutter_local_notifications/pull/1249.

Once that is merged in though, I'm going to look to have Linux support be in a pre-release instead to be safe given how it went last time and to let it "bake" for bit longer

**Comment by kennethnym** [comment]
I'm currently taking a crack at Windows implementation, however I am struggling with calling WinRT API in a win32 context. Calls to WinRT APIs always throws `HRESULT(ERROR_NOT_FOUND`, which after days of extensive research seems to indicate that the WinRT projection is not properly integrated (I think the program has to be bundled into an appx bundle). I would appreciate it if anyone can offer some guidance.

Related issues:
https://github.com/microsoft/microsoft-ui-xaml/issues/2773

**Comment by Sunbreak** [comment]
> I would appreciate it if anyone can offer some guidance.

Commented at https://github.com/MaikuB/flutter_local_notifications/issues/746#issuecomment-862322428

**Comment by kennethnym** [comment]
I set `VS_PROJECT_IMPORT` as well and I am still having issue.

**Comment by proninyaroslav** [comment]
@MaikuB 
Any plans to publish pre-release with the Linux support? I can't wait to test it out in real conditions :D

**Comment by Sunbreak** [comment]
> I set `VS_PROJECT_IMPORT` as well and I am still having issue.

@kennethnym We'd better discuss in another thread: [How to setup `WinRT/C++` for Flutter plugin on Windows](https://github.com/woodemi/quick_notify/discussions/8)

**Comment by bdlukaa** [comment]
https://pub.dev/packages/desktoasts works good on Windows. I could make a PR using it!

**Comment by clragon** [comment]
Well, it has been stated before that "`flutter_local_notifications` tries to avoid external dependencies",
so I suppose we could use [desktoasts](https://pub.dev/packages/desktoasts) as a template but we would need to implement our own version of all of it.

**Comment by kennethnym** [comment]
I completely forgot about this. I think I still have working notification on windows, I just need to integrate the logic to this library. I will try to come up with something next week.

**Comment by azchohfi** [comment]
Would a PR that uses the Windows App SDK be accepted? It is an external dependency, but it is the recommended way of uysing notifications on Windows. Full support should be available in 1.1 (https://github.com/microsoft/WindowsAppSDK#roadmap)
This should work in both un-packaged (a simple exe), or packaged (using the msix package) scenario, and it should also properly activate/open the app if it is closed, in both scenarios.

**Comment by kennethnym** [comment]
In my implementation I am using the winrt api that I pulled from nuget, but I don't think it counts as a third party dependency because it is from Microsoft themselves.

**Comment by azchohfi** [comment]
@kennethnym Alright! Let me know how I can help you. I want to see this landing as soon as possible. This is one of the most used Flutter dependencies, and I know that many apps would be unblocked by this work.

**Comment by MaikuB** [comment]
If it uses something official then sure. As an example, the Linux implementation uses the dbus package maintained by Canonical

**Comment by jinfagang** [comment]
Does windows notification avaiable now?

**Comment by kennethnym** [comment]
Not yet, still a WIP

**Comment by Sunbreak** [comment]
@kennethnym https://github.com/MaikuB/flutter_local_notifications/issues/746#issuecomment-890505993

> ```cmake
> set_target_properties(${PLUGIN_NAME} PROPERTIES
>   CXX_VISIBILITY_PRESET     hidden
>   VS_GLOBAL_TargetFramework "net5.0-windows10.0.17763.0"
>   VS_GLOBAL_CppWinRTPath    "${NUGET_PACKAGES_PATH}\\Microsoft.Windows.CppWinRT\\bin\\"
> )
> ```

Not sure what is your blocking issue. But `net5.0` is not necessary

BTW [WinRT Notification](https://docs.microsoft.com/en-us/uwp/api/windows.ui.notifications) is not available on Windows 7 if I record correctly

**Comment by kennethnym** [comment]
My visual studio install broke recently. I am trying to get it fixed.

**Comment by kennethnym** [comment]
Quick update: Flutter now bundles winrt so I no longer need to pull the winrt package manually from nuget. However, the bundled winrt seems to behave differently. I was able to send notifications successfully with the nuget version of winrt. WIth the bundled version of winrt, Windows shows "New notifications" as the content of the notification, instead of the body that I passed to it. I need to investigate further.

**Comment by azchohfi** [comment]
I believe it is not Flutter that does this, but the Windows SDK itself. There is a C++/WinRT version there, but you should always use the latest from nuget.org. I have a snippet that helps installing it on cmake:
https://github.com/Baseflow/flutter-geolocator/blob/726f928ac7662f417c69bb1958e4456afd173929/geolocator_windows/windows/CMakeLists.txt#L15-L45
You will probably also need these two compiler changes (to call async code, which most of the WinRT APIs do):
https://github.com/Baseflow/flutter-geolocator/blob/726f928ac7662f417c69bb1958e4456afd173929/geolocator_windows/windows/CMakeLists.txt#L54-L55
You might also need to link to the windowsapp lib, if you call any API from it:
```target_link_libraries(${PLUGIN_NAME} PRIVATE windowsapp)```

Let me know if this helps you.

**Comment by kennethnym** [comment]
That is very helpful. Thanks for the info!

**Comment by kennethnym** [comment]
Unfortunately, I am still having the same issue. If I use XML template to show notification, nothing would happen; if I use one of the default templates, only "New notifications" would appear.

**Comment by kennethnym** [comment]
I will make a draft PR.

**Comment by c0c4i** [comment]
Is it coming soon?

**Comment by kennethnym** [comment]
It will still take a while. Appreciate your patience!

**Comment by CoderUni** [comment]
Thank you for your work @kennethnym. I found a package that support windows notifications that you might be interested in: https://pub.dev/packages/local_notifier

**Comment by Fraa-124** [comment]
Hello, what is the state of this issue? With full support for desktop, this library is now a must for a lot of apps.

**Comment by kennethnym** [comment]
Hello, I'd love to close this as soon as possible, but I am still currently stuck at some issues (more in #1472). If someone could step in and help that would be great.

**Comment by ollydixon** [comment]
Looking forwards to the implementation.

**Comment by rekire** [comment]
What is the current status for the Windows support? What is still a problem which needs to be fixed?

**Comment by kmoreau** [comment]
Hello,
Is it possible to ask for the current status of the notifications on Windows? thanks :)

**Comment by prinkeeras** [comment]
Any updates on windows implementation?

**Comment by mattsrobot** [comment]
I would love to see Windows support!!

**Comment by realrk95** [comment]
Any updates on Windows?

**Comment by kennethnym** [comment]
sorry, i haven't worked with flutter in a long while, but anyone is free to pick up where i left off :)

**Comment by Levi-Lesches** [comment]
Thanks to @kennethnym and @lightrabbit, I'm currently working on #2349. Aiming to finish sometime this or next week.

**Comment by Levi-Lesches** [comment]
Update: Going really well so far. I've decided to support raw XML and Dart-based enums and classes for further support. Example in progress:

```dart
final action = WindowsNotificationAction(
  content: "Submit", 
  arguments: submit_picture",
);
final details = WindowsNotificationDetails(actions: [action]);
await plugin.show(_id++, "Title", "Body", notificationDetails: details);
```
![image](https://github.com/MaikuB/flutter_local_notifications/assets/20747538/0e26f7bc-ddd2-4fed-adf5-548a8ebcbd4a)

Clicking the button sends the string `submit_picture` back to Dart the usual way

**Comment by Levi-Lesches** [comment]
Update: I've transcribed the _entirety_ of the Windows Toast Notification API to Flutter. Here's a more complex example: 
```dart
    final profilePic = WindowsImage(
      source: File(r"C:\Users\KDAB\Downloads\image1.jpg"),
      altText: "The first image",
      crop: WindowsImageCrop.circle,
      placement: WindowsImagePlacement.appLogoOverride,
    );
    final bigImage = WindowsImage(
      source: File(r"C:\Users\KDAB\Downloads\image2.jpg"),
      altText: "The second image",
    );
    final likeButton = WindowsAction(
      content: "Like",
      arguments: "like_post|post_id",
    );
    final commentButton = WindowsAction(
      content: "View Comments",
      arguments: "comment_on_post|post_id",
    );
    const input = WindowsTextInput(
      id: "input-id",
      hintText: "Send a reply",
    );
    final sendButton = WindowsAction(
      content: "Send",
      arguments: "send_reply|post_id",
      inputId: "input-id",
    );
    final progress = WindowsProgressBar(status: "Syncing...", value: 0.75);
    final details = WindowsNotificationDetails(
      actions: [likeButton, commentButton, sendButton],
      images: [bigImage, profilePic],
      inputs: [input],
      progressBars: [progress],
    );
    await plugin.show(
      _id++,
      "Adam Wilson tagged you in a photo",
      "Isn't it great?",
      notificationDetails: details,
    );
```

![image](https://github.com/MaikuB/flutter_local_notifications/assets/20747538/59a6dc50-221d-4d34-ba01-91ce2d2b20ba)

Raw XML is also supported. I've updated #2349 and marked it as ready for review.

**Comment by Levi-Lesches** [comment]
Update: My branch is basically ready for use. Can't guarantee I won't break anything yet, but you can try out the `example/main.dart` file for yourself. On a non-MSIX, you can't search for or cancel specific notifications, and Windows doesn't support repeating notifications at all, but everything else (and a lot more) has been implemented in the example.

**Comment by Glitchy-Sheep** [comment]
@Levi-Lesches Wow, that's a great work, just awesome. Hope to see this in the next releases 💘

**Comment by Levi-Lesches** [comment]
Good people of this thread -- Windows support #2366 has been merged into `main`! Once the new release reaches Pub I expect we can close this issue
